src/main/java/use_cases/PlayerGameInteractor.java

```diff
@@ -0,0 +1,212 @@
+        }  // The card is not valid, therefore it should not be played, nothing is to be done.
+
+    }
+
+    /**
+     * Implement the logic required when a User requests to pick up a Card, that is, check if the user
+     * can pick up a Card, and if they can, then pick up one, otherwise don't do anything.
+     */
+    private void pickUpCardRequestLogic(Player currentPlayer) {
+        if(!(currentGame.getCurrentTurnHasPickedUp()) && !(anyValidCards(currentPlayer))) {
+            currentPlayer.pickUpCard(currentGame);
+            currentGame.notifyGameObservers();
+        }
+    }
+
+    /**
+     * Implement the logic required when a User requests to skip a turn, that is to check if the user
+     * has any valid cards or if the user has not picked up, if they do then do not skip, otherwise skip
+     */
+    private void skipTurnLogic(Player currentPlayer) {
+        assert currentPlayer.equals(currentGame.getCurrentTurn());
+        if ((currentGame.getCurrentTurnHasPickedUp()) & !(anyValidCards(currentPlayer))) {
+            // User can skip.
+            currentGame.changeCurrentTurn();
+            currentGame.notifyGameObservers();
+        }
+            // User cannot skip.
+
+        }
+
+    /**
+     * Check all the Cards in this Player's Hand to see if any of them are valid to play in the Game.
+     * @return True iff any Cards are valid to play, false otherwise.
+     */
+    private boolean anyValidCards(Player player) {
+        for (Card card : player.getCards()) {
+            if (currentGame.isValidCard(card)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    /**
+     * Perform the logic required when a Player wins the Game, that is, set the Player as the Game winner,
+     * increment wins and losses for all Players
+     * @param player The player who has won the Game.
+     */
+    private void winLogic(Player player) {
+        currentGame.setWinner(player);
+        if(player instanceof HumanPlayer) {
+            ((HumanPlayer) player).incrementWins();
+        }
+        for(Player p: currentGame.getPlayers()) {
+            if(p instanceof HumanPlayer && !(p.equals(player))) {
+                ((HumanPlayer) p).incrementLosses();
+            }
+        }
+        currentGame.notifyGameObservers();
+    }
+
+    /**
+     * Check if the given Player is the winner of the Game
+     * @return True iff the Player has no Cards left in their Hand (played their last Card).
+     */
+    private boolean isWinner(Player player) {
+        return player.getCards().isEmpty();
+    }
+
+
+    /**
+     * This method implements all the logic for ComputerPlayers, including the logic for putting down a Card,
+     * picking one up, and skipping the turn.
+     * @param compPlayer The ComputerPlayer whose turn it is.
+     */
+    private void computerPlayerLogic(ComputerPlayer compPlayer) {
+        Card compCard = compPlayer.selectRandomCard(this.currentGame);
```

Comment:
Consider renaming to selectRandomValidCard? That way the name of the method better explains what the code actually does, rather than selecting a card completely at random.

Reply:
Done